### PR TITLE
feat(net): return peer metadata from `connect_isolated` functions

### LINKED ISSUE #4678: Return peer metadata from `connect_isolated` functions
## Motivation

We want to support development of an alternate DNS seeder based on `zebra-network`.

## Requirements

There's some peer metadata which we need access to (from the `connect_isolated`-style functions) in order to reproduce the [seeder filtering algorithms](https://github.com/zcash/zcash/issues/3881#issuecomment-1145555036) and I'm unsure how to make that happen:

* The value of the `PeerServices` bitset
* The numeric protocol version number (the stuff that looks like `170100`)
* The textual user-agent string (the stuff that looks like `/MagicBean:5.0.0/`)

I see there's a `negotiate_version` in `handshake.rs` that seems to do the version handshaking and it returns a tuple `(remote_version, remote_services, remote_canonical_addr)` (which is almost all we want, it's missing the text user-agent string).

The place where `negotiate_version` is called is `impl<S, PeerTransport, C> Service<HandshakeRequest<PeerTransport>> for Handshake<S, C>`; but I'm not sure how the latter gets called exactly, and it doesn't return/pass through the remote version/services info.

_Originally posted by @superbaud in https://github.com/zcash/zcash/issues/3881#issuecomment-1162096689_

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to support development of an alternate DNS seeder based on `zebra-network`.

Depends-On: #4869

Closes #4678

### Designs

See the detailed design in ticket #4678.

## Solution

Interface changes:
- Return a Client instance from connect_isolated_* functions
- Export the Client type from zebra-network
- Add a copy of the Version message to the ConnectionInfo struct
- Export the types required to use the new APIs

Refactors & improvements:
- Move LoadTrackedCient version into a ConnectionInfo struct
- Add negotiated version to ConnectionInfo
- Add the peer address to ConnectionInfo, add ConnectionInfo to Connection
- Add peer services and user agent to ConnectionInfo

Related fixes:
- Add and improve debug formatting

## Review

Zebra developers should feel free to review this, but we need to check if it works for the zebra-network users first.

### Reviewer Checklist

  - [ ] Code implements Design
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Tasks

- Add a copy of the `ConnectionInfo` to `MetaAddr`
- Write tests that make sure remote peer metadata gets turned into the correct `ConnectionInfo` fields
    - peer set connections
    - isolated connections

### CODE REVIEW

**Comment by softminus** [comment]
(I'm writing a seeder in Rust that uses the `zebra_network` crate)

My bad for not realizing we'd need this as well, but we definitely need access to `start_height` (the current C++ seeder logic uses that as one of the signals to classify peers) as well and the `relay` field too (the current seeder doesn't use it but it might be useful) from the `Version` message: https://doc-internal.zebra.zfnd.org/zebra_network/protocol/external/message/enum.Message.html#variant.Version

I'm wondering if it'd be possible to just include a deserialized copy of the `Version` message we received from the other end?

But besides that it seems OK, I'm going to try seeing if I can aim my `Cargo.toml` at your forked branch and get my in-progress seeder code working with it!

**Comment by softminus** [comment]
Also I think we need to make all the types for the fields of `connection_info` accessible from the outside, when I do `use zebra_network::protocol::external::types::Version;` i get:
```
sasha_work@seasonal-dream zcash-cotyledon % cargo run
   Compiling zcash-cotyledon v0.1.0 (/Users/sasha_work/remote_vm_repos/zcash-cotyledon)
error[E0603]: module `protocol` is private
   --> src/main.rs:20:20
    |
20  | use zebra_network::protocol::external::types::Version;
    |                    ^^^^^^^^ private module
    |
note: the module `protocol` is defined here
   --> /Users/sasha_work/repos/zebra/zebra-network/src/lib.rs:153:1
    |
153 | mod protocol;
    | ^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0603`.
error: could not compile `zcash-cotyledon` due to previous error
sasha_work@seasonal-dream zcash-cotyledon % 
```

and likewise for `connected_addr` and `peer_services`: 
```
sasha_work@seasonal-dream zcash-cotyledon % cargo run
   Compiling zcash-cotyledon v0.1.0 (/Users/sasha_work/remote_vm_repos/zcash-cotyledon)
error[E0603]: module `peer` is private
   --> src/main.rs:10:21
    |
10  | use  zebra_network::peer::handshake;
    |                     ^^^^ private module
    |
note: the module `peer` is defined here
   --> /Users/sasha_work/repos/zebra/zebra-network/src/lib.rs:150:1
    |
150 | mod peer;
    | ^^^^^^^^^

error[E0603]: module `protocol` is private
   --> src/main.rs:11:20
    |
11  | use zebra_network::protocol::external::types;
    |                    ^^^^^^^^ private module
    |
note: the module `protocol` is defined here
   --> /Users/sasha_work/repos/zebra/zebra-network/src/lib.rs:153:1
    |
153 | mod protocol;
    | ^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0603`.
error: could not compile `zcash-cotyledon` due to 2 previous errors
sasha_work@seasonal-dream zcash-cotyledon % 
```

**Comment by teor2345** [comment]
> My bad for not realizing we'd need this as well, but we definitely need access to `start_height` (the current C++ seeder logic uses that as one of the signals to classify peers)

Oh, that's probably the cause of Zebra bug #1904 (which is unrelated to what you're doing, but easy for us to fix).

> as well and the `relay` field too (the current seeder doesn't use it but it might be useful) from the `Version` message: https://doc-internal.zebra.zfnd.org/zebra_network/protocol/external/message/enum.Message.html#variant.Version
> 
> I'm wondering if it'd be possible to just include a deserialized copy of the `Version` message we received from the other end?

Yep, that's probably fine, we'll wrap it in an `Arc` to save memory in the address book. (Otherwise we'll have about 5 copies of each Version message.)

> But besides that it seems OK, I'm going to try seeing if I can aim my `Cargo.toml` at your forked branch and get my in-progress seeder code working with it!

Great!

**Comment by teor2345** [comment]
> Also I think we need to make all the types for the fields of `connection_info` accessible from the outside

Some of those types might be exposed at a higher module level. But if not, that's an easy fix, feel free to submit a PR, or I'll get to it later this week.

**Comment by softminus** [comment]
> > My bad for not realizing we'd need this as well, but we definitely need access to `start_height` (the current C++ seeder logic uses that as one of the signals to classify peers)
> 
> Oh, that's probably the cause of Zebra bug #1904 (which is unrelated to what you're doing, but easy for us to fix).

I think that that code is commented out in the version that's actually running:
```
static inline int GetRequireHeight(const bool testnet = fTestNet)
{
    // return testnet ? 500000 : 350000;
    return 0;
}
```
and I'm not quite sure why, but I think that makes it not the cause.

If it'd help, I definitely could try aiming the extant C++ seeder at a zebra node and seeing why it's not showing up. Do you have an IP address / DNS name of a long-running zebra node I could use for that?

**Comment by teor2345** [comment]
> If it'd help, I definitely could try aiming the extant C++ seeder at a zebra node and seeing why it's not showing up. Do you have an IP address / DNS name of a long-running zebra node I could use for that?

That would be great!

We can ask around for one, all the ones I know about are transient IPs, nodes, or firewalled. Or @gustavovalverde will have a long-term seeder IP once we fix deployments in tickets #4546 and #4562.

I'll leave it to you both to work that out - we can open a Zebra ticket for the testing if it would help.

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
> > as well and the `relay` field too (the current seeder doesn't use it but it might be useful) from the `Version` message: https://doc-internal.zebra.zfnd.org/zebra_network/protocol/external/message/enum.Message.html#variant.Version
> > I'm wondering if it'd be possible to just include a deserialized copy of the `Version` message we received from the other end?
> 
> Yep, that's probably fine, we'll wrap it in an `Arc` to save memory in the address book. (Otherwise we'll have about 5 copies of each Version message.)

I split out a `VersionMessage` struct, and added it to the `Client` struct.

I also fixed up the `zebra_network` exports, and added a test to make sure that any future exports happen.

**Comment by teor2345** [comment]
@superbaud just wondering what you thought of these new changes?

**Comment by superbaud** [comment]
> @superbaud just wondering what you thought of these new changes?

I'm gonna take a look at them today and tomorrow, sorry for the delay!

**Comment by softminus** [comment]
> @superbaud just wondering what you thought of these new changes?

I'm gonna take a look at them today and tomorrow, sorry for the delay and thanks for working on this!

**Comment by teor2345** [comment]
Oops, I made a mistake with the rebase, here's the correct one.

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
Let's merge this new code to avoid code conflicts. If we need to make more changes to support DNS seeders, we can do them later.

**Comment by teor2345** [comment]
@mergifyio update

File: zebra-network/src/peer/handshake.rs [hunk 1]
```diff
@@ -98,6 +117,26 @@ where
...
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ConnectionInfo {
+    /// The connected peer address, if known.
+    /// This address might not be valid for outbound connections.
+    ///
+    /// Peers can be connected via a transient inbound or proxy address,
+    /// which will appear as the connected address to the OS and Zebra.
+    pub connected_addr: ConnectedAddr,
+
+    /// The network protocol [`Message::Version`](crate::Message) sent by the remote peer.
```
Comment (teor2345) [inline zebra-network/src/peer/handshake.rs]: Fix this doc link

File: zebra-network/src/lib.rs [hunk 2]
```diff
@@ -168,15 +168,24 @@ pub use crate::{
...
     policies::RetryLimit,
-    protocol::internal::{InventoryResponse, Request, Response},
+    protocol::{
+        external::{Version, VersionMessage},
+        internal::{InventoryResponse, Request, Response},
+    },
 };
 
-/// Types used in the definition of [`Request`] and [`Response`] messages.
+/// Types used in the definition of [`Request`], [`Response`], and [`Message::Version`] messages.
```
Comment (teor2345) [inline zebra-network/src/lib.rs]: Fix link

File: zebra-network/src/peer/handshake.rs (hunk 1 above)
Comment (teor2345) [reply zebra-network/src/peer/handshake.rs]: Fixed in df9b50b.

File: zebra-network/src/lib.rs (hunk 2 above)
Comment (teor2345) [reply zebra-network/src/lib.rs]: Fixed in df9b50b.

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**